contrib/vdsm-dev

```diff
@@ -0,0 +1,187 @@
+}
+
+venv_create()
+{
+    echo "Creating venv at $venv_path"
+    python3 -m venv ${venv_path}
+    source "${venv_activate_path}"
+    pip install --upgrade pip
+    pip install -r docker/requirements.txt
+    deactivate
+}
+
+make_storage()
+{
+    if [ ! -d $user_storage_path ]; then
+        echo "Make storage at $user_storage_path"
+        make storage
+    fi
+}
+
+test_path()
+{
+    path=$1
+    storage_prefix=storage/
+    if [ ! -f "tests/$path" ]; then
+        path=${storage_prefix}${path}
+        if [ ! -f "tests/$path" ]; then
+            echo "No test $path found"
+            exit 1
+        fi
+    fi
+}
+
+vdsm_build()
+{
+    git clean -dxf && ./autogen.sh --system --enable-timestamp && \
+         make && rm -rf ~/rpmbuild && make rpm
+}
+
+handle_test_opts()
+{
+    while test $# -gt 0
+    do
+        case "$1" in
+            -e)
+                shift
+                env=$1
+                if [ $env == "storage" ]; then
+                    make_storage
+                fi
+                ;;
+            *)
+                # Let tox handle the rest of the arguments
+                ;;
+        esac
+        shift
+    done
+}
+
+
+if [ -z "$1" ]; then
+    echo "Command missing"
+    print_help
+    exit 1
+fi
+
+case $1 in
+
+    help )
+
+        print_help
+
+        ;;
+
+    build )
+
+        vdsm_build
+
+        ;;
+
+    build-install )
+
+        vdsm_build && (cd ~/rpmbuild/RPMS && sudo dnf upgrade */vdsm*.rpm)
+
+        ;;
+
+    test )
+
+        if [ ! -f $venv_activate_path ]; then
+            venv_create
+        fi
+
+        handle_test_opts "$@"
+
+        source ${venv_activate_path}
+        echo "Running: tox ${*:2}"
+        tox "${*:2}"
+
+        ;;
+
+    test-storage )
+
+        shift
+        test=$1
+
+        if [[ "$test" == *":"* ]]; then
+            IFS=':' read -a filepath rest <<<"$test"
+            test_path $filepath
+        else
+            test_path $test
+        fi
+        
+        if [ ! -f $venv_activate_path ]; then
+            venv_create
+        fi
+
+        make_storage
+
+        source ${venv_activate_path}
+        if [[ $test == "" ]]; then
+            echo "Running: tox -e storage"
+            tox -e storage
+        else
+            echo "Running: tox -e storage -- -vv $test ${*:2}"
+            tox -e storage -- -vv $test "${*:2}"
+        fi
+
+        ;;
+
+    show-coverage )
+
+        env="storage"
+        while test $# -gt 0
+        do
+            case "$1" in
+                -e)
+                    shift
+                    env=$1
+                    ;;
+                *)
+                    echo "Unrecognized parameter: $1"
+                    ;;
+            esac
+            shift
+        done
+
+        cov_path="tests/htmlcov-$env/index.html"
+
+        if [ ! -f $cov_path ]; then
+            echo "No coverage summary found at $cov_path"
+            exit 1
+        fi
+
+        xdg-open $cov_path
```

Comment:
This can also be nice as make target, maybe:

    make storage-coverage

Reply:
If it is only for storage, yes, it might be nice to have a make target for that.

In this script, it works nicely for any environment. Doing the same in a Makefile is not possible (well, it is by hacking the makefile and receiving envvars), or it would result in a bunch of targets for all test environments.

Reply:
For coverage, it is useful to show coverage per tox env, or
combined coverage for the entire code. We already have code
to generate the combined coverage. I'm not sure how important
is the coverage per tox env, make just add

   make coverage

Creating the combined coverage and opening the index.html file in a browser.

We are also missing a way to publish coverage report from github builds.
I think the best way would be to integrate with services like [coveralls](https://coveralls.io/).

Reply:
Where is the combined coverage generated? Is it the `lib` environment?

Reply:
The code is in automation/check-patch*.sh. We used to run this after all the tests and publish the results in jenkins.